### PR TITLE
Video Editor: Add new video-editor block to enable selecting a custom video poster

### PR DESCRIPTION
**Pull request by donnapep**
This PR separates out the video editor block from PR #11389, which adds support for selecting or uploading a custom image to serve as the poster for a video.

P2 post discussing Calypso-based VideoPress poster chooser can be found at pxWta-JO-p2 (internal ref).

### CODE REVIEW

File: client/blocks/video-editor/video-editor-buttons.jsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+					onClick={ onUploadImageClick }
+					onUploadImage={ onUploadImage }>
+					{ translate( 'Upload Image' ) }
+				</UploadButton>
+				<Button
+					className="video-editor__buttons-button"
+					disabled={ ! isVideoLoaded || isPosterUpdating }
+					onClick={ onSelectFrame }
+					primary>
+					{ translate( 'Select Frame' ) }
```
Comment (a8ci18n) [inline client/blocks/video-editor/video-editor-buttons.jsx]: Hi! I've found a possible matching string that has already been translated **28** times:
`translate( 'Frames' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by donnapep** [comment]
@roccotripaldi Thanks so much for your review of this. Much appreciated! I hope to be able to get back to this soon.

**Comment by donnapep** [comment]
@gwwar Would appreciate an extra set of eyes on this since you worked on the image editor. Thx!

**Comment by gwwar** [comment]
👍 I can this this for a spin tomorrow

**Comment by gwwar** [comment]
@donnapep Do you mind adding some testing instructions? 

cc @Automattic/stark as well

**Comment by donnapep** [comment]
@gwwar Hmmm...there's not really anything to test in Calypso since the block isn't being used anywhere yet. That will happen in PR #11653, so I think the bulk of the testing will happen there. I've been testing locally by just copying the changes from this PR into my local branch of that other PR.

**Comment by gwwar** [comment]
@donnapep for new blocks, there's always the option to add them to add an example to dev docs. https://wpcalypso.wordpress.com/devdocs/blocks

File: client/blocks/video-editor/index.jsx [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+		return {
+			posterUrl: getPosterUrl( state ),
+			shouldCloseModal: shouldCloseVideoEditorModal( state ),
+			shouldShowError: shouldShowVideoEditorError( state ),
+			uploadProgress: getPosterUploadProgress( state ),
+		};
+	},
+	dispatch => bindActionCreators( {
+		updatePoster,
+	}, dispatch ),
```
Comment (gwwar) [inline client/blocks/video-editor/index.jsx]: If we're not doing anything custom, mapDispatchToProps can also be an object. This can be shortened to `{ updatePoster }`. See also https://github.com/reactjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options

File: client/blocks/video-editor/index.jsx [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+				pauseVideo: false
+			} );
+
+			return;
+		}
+
+		this.props.onUpdatePoster( this.getVideoEditorProps( nextProps.posterUrl ) );
+	}
+
+	handleSelectFrame = () => {
```
Comment (gwwar) [inline client/blocks/video-editor/index.jsx]: What do we think of the naming scheme here? This is optional, but I think we can safely drop the`handle` prefix on these functions without losing context

File: client/blocks/video-editor/index.jsx [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+			return;
+		}
+
+		isSelectingFrame = true;
+
+		this.setState( {
+			error: false,
+			pauseVideo: true
+		} );
+	}
```
Comment (gwwar) [inline client/blocks/video-editor/index.jsx]: We don't have a hard rule on this, but I prefer to add a `;` here since this is an assignment.

File: client/blocks/video-editor/index.jsx [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+/**
+ * Module variables
+ */
+let isSelectingFrame = false;
+
+class VideoEditor extends Component {
+	static propTypes = {
+		className: PropTypes.string,
+		media: PropTypes.object.isRequired,
```
Comment (gwwar) [inline client/blocks/video-editor/index.jsx]: Also optional, but the component may be easier to use if we can handle media not being available initially.

File: client/blocks/video-editor/index.jsx [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+		if ( nextProps.shouldShowError ) {
+			this.setState( {
+				error: true,
+				pauseVideo: false
+			} );
+
+			return;
+		}
+
+		this.props.onUpdatePoster( this.getVideoEditorProps( nextProps.posterUrl ) );
```
Comment (gwwar) [inline client/blocks/video-editor/index.jsx]: Let's also check to see if posterUrl has changed. Eg
```jsx
if ( this.props.posterUrl !== nextProps.posterUrl ) {
```

File: client/blocks/video-editor/index.jsx [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+		isLoading: true,
+		pauseVideo: false,
+	};
+
+	componentWillReceiveProps( nextProps ) {
+		if ( ! nextProps.shouldCloseModal && ! nextProps.shouldShowError ) {
+			return;
+		}
+
+		if ( nextProps.shouldShowError ) {
```
Comment (gwwar) [inline client/blocks/video-editor/index.jsx]: Same thing here:
```jsx
if ( nextProps.shouldShowError &&  ! this.props.shouldShowError ) {
```

As a aside, why do we need to store error state, if we have this passed as a prop?

File: client/blocks/video-editor/index.jsx [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+	getPosterUploadProgress,
+	getPosterUrl,
+	shouldCloseVideoEditorModal,
+	shouldShowVideoEditorError,
+} from 'state/selectors';
+
+/**
+ * Module variables
+ */
+let isSelectingFrame = false;
```
Comment (gwwar) [inline client/blocks/video-editor/index.jsx]: Why is this here? Can we not store this in component or global state?

File: client/blocks/video-editor/index.jsx [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+		} );
+	}
+
+	handlePause = ( currentTime ) => {
+		if ( ! isSelectingFrame ) {
+			return;
+		}
+
+		const { media } = this.props;
+		const guid = media && media.videopress_guid ? media.videopress_guid : null;
```
Comment (gwwar) [inline client/blocks/video-editor/index.jsx]: I would recommend lodash `get` here:
```jsx
const guid = get( media, 'videopress_guid', null );
```

File: client/blocks/video-editor/index.jsx [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+			pauseVideo: true
+		} );
+	}
+
+	handlePause = ( currentTime ) => {
+		if ( ! isSelectingFrame ) {
+			return;
+		}
+
+		const { media } = this.props;
```
Comment (gwwar) [inline client/blocks/video-editor/index.jsx]: We can also pull out updatePoster, since we're already destructuring here.

```
const { updatePoster, media } = this.props;
```

File: client/blocks/video-editor/index.jsx [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+		const guid = media && media.videopress_guid ? media.videopress_guid : null;
+
+		if ( guid ) {
+			this.props.updatePoster( guid, { atTime: Math.floor( currentTime ) } );
```
Comment (gwwar) [inline client/blocks/video-editor/index.jsx]: What format is currentTime, and why do we need to floor this? JSdoc for this function may be helpful in this case.

File: client/blocks/video-editor/index.jsx [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+	handleUploadImageClick = () => {
+		isSelectingFrame = false;
+
+		this.setState( {
+			error: false,
+			pauseVideo: true
+		} );
+	}
+
+	handleUploadImage = ( file ) => {
```
Comment (gwwar) [inline client/blocks/video-editor/index.jsx]: Maybe some jsdoc? What is a file here?

File: client/blocks/video-editor/style.scss [hunk 13]
```diff
@@ -0,0 +1,90 @@
+.video-editor {
+	position: relative;
+	width: 100%;
+	height: 100%;
+}
+
+.editor-media-modal .video-editor .notice {
```
Comment (gwwar) [inline client/blocks/video-editor/style.scss]: Let's pass a classname to the notice, maybe something like `.video-editor__error-notice`. This will let us simplify the selector down to:
```
.video-editor__error-notice {
```

File: client/blocks/video-editor/video-editor-buttons.jsx [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+	onSelectFrame,
+	onUploadImage,
+	onUploadImageClick,
+	translate,
+} ) => {
+	return (
+		<div className="video-editor__buttons">
+			{ onCancel &&
+				<Button
+					className="video-editor__buttons-button"
```
Comment (gwwar) [inline client/blocks/video-editor/video-editor-buttons.jsx]: Perhaps rename the component VideoEditorControls, so this can have a less buttony name 😄 eg something like `video-editor__controls-button`

File: client/blocks/video-editor/style.scss [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+		top: 50%;
+		left: 50%;
+	height: 100%;
+	width: 100%;
+	max-height: 100%;
+	max-width: 100%;
+	transform: translate( -50%, -50% );
+}
+
+.video-editor__content .progress-bar {
```
Comment (gwwar) [inline client/blocks/video-editor/style.scss]: I believe we can also pass a classname to the progress bar.

File: client/blocks/video-editor/video-editor-upload-button.js [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+		onClick: noop,
+		onUploadImage: noop,
+	};
+
+	setFormInstance = ref => this.form = ref;
+
+	handleChange = ( event ) => {
+		let files;
+
+		if ( event.target.files && event.target.files.length > 0 ) {
```
Comment (gwwar) [inline client/blocks/video-editor/video-editor-upload-button.js]: A lodash get may be helpful here;

```jsx
const files = get( event, 'target.files[0]' );

File: client/blocks/video-editor/video-editor-upload-button.js [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+			files = event.target.files[ 0 ];
+		}
+
+		this.props.onUploadImage( files );
```
Comment (gwwar) [inline client/blocks/video-editor/video-editor-upload-button.js]: If files is falsey, should we still call this handler?

File: client/blocks/video-editor/video-editor-upload-button.js [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+		ReactDom.findDOMNode( this.form ).reset();
```
Comment (gwwar) [inline client/blocks/video-editor/video-editor-upload-button.js]: Can we think of a way to avoid this? Perhaps we can control the `value`prop of the input, or add a key prop to force a new instance to render:
```
<input
    key={ this.state.key }
```

File: client/blocks/video-editor/video-editor-upload-button.js [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+
+	static defaultProps = {
+		isPosterUpdating: false,
+		onClick: noop,
+		onUploadImage: noop,
+	};
+
+	setFormInstance = ref => this.form = ref;
+
+	handleChange = ( event ) => {
```
Comment (gwwar) [inline client/blocks/video-editor/video-editor-upload-button.js]: Maybe `fileSelected`?

**Comment by gwwar** [comment]
Pretty cool so far @donnapep. I left a number of notes, many of them being optional, but I'd really like us to move `isSelectingFrame` to component state or another place that's more appropriate.

Also, would it be possible to stub out the example a bit further? It's a bit unclear what should happen next.
screen shot 2017-06-08 at 12 39 18 pm

File: client/blocks/video-editor/index.jsx (hunk 2 above)
Comment (donnapep) [inline client/blocks/video-editor/index.jsx]: Fixed in https://github.com/Automattic/wp-calypso/pull/11648/commits/b798412db227139ac9f05646a39b87d018d482b1.

File: client/blocks/video-editor/index.jsx (hunk 3 above)
Comment (donnapep) [inline client/blocks/video-editor/index.jsx]: Updated in https://github.com/Automattic/wp-calypso/pull/11648/commits/c57e3e33d281d3d03907559506474aa3d55b62ca.

File: client/blocks/video-editor/index.jsx (hunk 6 above)
Comment (donnapep) [inline client/blocks/video-editor/index.jsx]: I did some refactoring of this method in https://github.com/Automattic/wp-calypso/pull/11648/commits/79fd488bb84b2a9223f701761ccd45025a031af0 and realized I don't actually need a separate action for determining if the modal should be closed or not.

File: client/blocks/video-editor/index.jsx (hunk 7 above)
Comment (donnapep) [inline client/blocks/video-editor/index.jsx]: That's because `shouldShowError` only indicates a problem with trying to update the poster, but there are other types of errors (e.g. [error loading the VideoPress script](https://github.com/Automattic/wp-calypso/pull/11648/files#diff-35b626cb9b7af72412a7ab302d8fc3f6R185)). They also result in [rendering the same error](https://github.com/Automattic/wp-calypso/pull/11648/files#diff-35b626cb9b7af72412a7ab302d8fc3f6R138) to the user though.

Do you think the naming could be improved somewhere to reduce confusion?

File: client/blocks/video-editor/index.jsx (hunk 8 above)
Comment (donnapep) [inline client/blocks/video-editor/index.jsx]: Yeah, that's strange. Fixed in https://github.com/Automattic/wp-calypso/pull/11648/commits/3e417f95f316a8ce9c4353ea744cd2f1af99b194.

File: client/blocks/video-editor/index.jsx (hunk 9 above)
Comment (donnapep) [inline client/blocks/video-editor/index.jsx]: Fixed by https://github.com/Automattic/wp-calypso/pull/11648/commits/e8c3ca9c6e984e41b64aff0173cd85af0abdad93.

File: client/blocks/video-editor/index.jsx (hunk 10 above)
Comment (donnapep) [inline client/blocks/video-editor/index.jsx]: There's already a function in the component named `updatePoster`, so it doesn't seem like I can do that.

Comment (gwwar) [inline client/blocks/video-editor/index.jsx]: This was optional (I often use this.props for short functions), but if you need a new variable name that's possible with destructure too.
```jsx
const { updatePoster: myNewName, media } = this.props;
```

File: client/blocks/video-editor/index.jsx (hunk 11 above)
Comment (donnapep) [inline client/blocks/video-editor/index.jsx]: I removed the `Math.floor` in [this commit](https://github.com/Automattic/wp-calypso/pull/11648/commits/797a870e1203a75821ebb4861a075a388392108e). I did some testing w/o it and couldn't see any reason for it to be there. 🤷‍♀️

Comment (gwwar) [inline client/blocks/video-editor/index.jsx]: If currentTime is in ms then`Math.floor( currentTime )` would be the same as `currentTime`.

If you were aiming to round to the nearest second, but keep the units in ms, that would be something like:
```
Math.round( currentTime / 1000 ) * 1000
```

File: client/blocks/video-editor/index.jsx (hunk 12 above)
Comment (donnapep) [inline client/blocks/video-editor/index.jsx]: Clarified in https://github.com/Automattic/wp-calypso/pull/11648/commits/40ddb83be3e9a3d36966cb4111063aa647571b97.

File: client/blocks/video-editor/style.scss (hunk 13 above)
Comment (donnapep) [inline client/blocks/video-editor/style.scss]: I think it needs to be this way in order to be able to beat [this rule](https://github.com/Automattic/wp-calypso/blob/master/client/post-editor/media-modal/index.scss#L139).
